File: kernel/c/root-linux/src/dbsimple.c
/***********************************************************************
 *          dbsimple.c
 *
 *          Simple DB file for persistent attributes
 *
 *          Copyright (c) 2015 Niyamaka.
 *          Copyright (c) 2024, ArtGins.
 *          All Rights Reserved.
***********************************************************************/
#include <string.h>
#include <unistd.h>
#include <limits.h>

#include <kwid.h>
#include "yunetas_environment.h"
#include "dbsimple.h"

/***************************************************************
 *              Constants
 ***************************************************************/

/***************************************************************
 *              Structures
 ***************************************************************/

/***************************************************************
 *              Prototypes
 ***************************************************************/

/***************************************************************
 *              Data
 ***************************************************************/

/***************************************************************************
 *
 ***************************************************************************/
PRIVATE char *get_persist_filename(
    hgobj gobj,
    char *bf,
    int bflen,
    const char *label,
    BOOL create_directories)
{
    char filename[NAME_MAX];
    snprintf(filename, sizeof(filename), "%s-%s-%s.json",
        gobj_gclass_name(gobj),
        gobj_name(gobj),
        label
    );

    return yuneta_realm_file(bf, bflen, "data", filename, create_directories);
}

/***************************************************************************
 *
 ***************************************************************************/
PRIVATE json_t *load_json(
    hgobj gobj
)
{
    char filename[PATH_MAX];
    get_persist_filename(gobj, filename, sizeof(filename), "persistent-attrs", false);

    if(!is_regular_file(filename)) {
        // No persistent attrs saved
        return 0;
    }

    size_t flags = 0;
    json_error_t error;
    json_t *jn_device = json_load_file(filename, flags, &error);
    return jn_device;
}

/***************************************************************************
 *
 ***************************************************************************/
PRIVATE int save_json(
    hgobj gobj,
    json_t *jn // owned
)
{
    char filename[NAME_MAX];
    get_persist_filename(gobj, filename, sizeof(filename), "persistent-attrs", true);

    int ret = json_dump_file(
        jn,
        filename,
        JSON_INDENT(4)
    );
    if(ret < 0) {
        gobj_log_error(gobj, 0,
            "function",     "%s", __FUNCTION__,
            "msgset",       "%s", MSGSET_JSON_ERROR,
            "msg",          "%s", "Cannot save device json database",
            "path",         "%s", filename,
            NULL
        );
        JSON_DECREF(jn)
        return -1;
    }
    JSON_DECREF(jn)
    return 0;
}

/***************************************************************************
   Load writable persistent attributes from simple file db
 ***************************************************************************/
PUBLIC int db_load_persistent_attrs(
    hgobj gobj,
    json_t *keys  // owned
) {
    json_t *jn_file = load_json(gobj);
    if(jn_file) {
        json_t *attrs = kw_clone_by_keys(
            gobj,
            jn_file,    // owned
            keys,       // owned
            false
        );

        gobj_write_attrs(gobj, attrs, SDF_PERSIST, 0);
    }

    return 0;
}

/***************************************************************************
 *
 ***************************************************************************/
PUBLIC int db_save_persistent_attrs(
    hgobj gobj,
    json_t *keys  // owned
) {
    json_t *jn_attrs = gobj_read_attrs(gobj, SDF_PERSIST, 0);
    json_t *attrs = kw_clone_by_keys(
        gobj,
        jn_attrs,   // owned
        keys,       // owned
        false
    );

    json_t *jn_file = load_json(gobj);
    if(jn_file) {
        json_object_update_missing(attrs, jn_file);
        JSON_DECREF(jn_file)
    }

    save_json(
        gobj,
        attrs  // owned
    );

    return 0;
}

/***************************************************************************
 *
 ***************************************************************************/
PUBLIC int db_remove_persistent_attrs(
    hgobj gobj,
    json_t *keys  // owned
) {
    json_t *jn_file = load_json(gobj);

    json_t *attrs = kw_clone_by_not_keys(
        gobj,
        jn_file,    // owned
        keys,       // owned
        false
    );

    save_json(
        gobj,
        attrs  // owned
    );

    return 0;
}

/***************************************************************************
 *
 ***************************************************************************/
PUBLIC json_t *db_list_persistent_attrs(
    hgobj gobj,
    json_t *keys  // owned
) {
    json_t *jn_file = load_json(gobj);

    json_t *attrs = kw_clone_by_keys(
        gobj,
        jn_file,    // owned
        keys,       // owned
        false
    );
    return attrs;
}

File: kernel/js/yunetas-js7/yunetas/src/dbsimple.js
/*********************************************************************************
 *      dbsimple.js
 *
 *      Licence: MIT (http://www.opensource.org/licenses/mit-license)
 *      Copyright (c) 2014,2024 Niyamaka.
 *      Copyright (c) 2025, ArtGins.
 *********************************************************************************/

import {
    kw_get_local_storage_value,
    kw_remove_local_storage_value,
    kw_set_local_storage_value,
    is_object,
    kw_clone_by_keys,
    json_object_update_missing,
} from "./helpers.js";

import {
    sdata_flag_t,
    gobj_read_attrs,
    gobj_short_name,
    gobj_write_attrs,
} from "./gobj.js";

/************************************************************
 *
 ************************************************************/
function _get_persistent_path(gobj)
{
    return "persistent-attrs-" + gobj_short_name(gobj);
}

/************************************************************
 *
 ************************************************************/
function db_load_persistent_attrs(
    gobj,
    keys  // str, list or dict.
)
{
    let jn_file = kw_get_local_storage_value(_get_persistent_path(gobj), null, false);
    if(jn_file && is_object(jn_file)) {
        let attrs = kw_clone_by_keys(
            gobj,
            jn_file,    // owned
            keys,       // owned
            false
        );

        gobj_write_attrs(gobj, attrs, sdata_flag_t.SDF_PERSIST, 0);
    }
}

/************************************************************
 *
 ************************************************************/
function db_save_persistent_attrs(
    gobj,
    keys  // str, list or dict.
)
{
    let jn_attrs = gobj_read_attrs(gobj, sdata_flag_t.SDF_PERSIST, 0);
    let attrs = kw_clone_by_keys(
        gobj,
        jn_attrs,   // owned
        keys,       // owned
        false
    );

    let jn_file = kw_get_local_storage_value(_get_persistent_path(gobj), null, false);
    if(jn_file && is_object(jn_file)) {
        json_object_update_missing(attrs, jn_file);
    }

    kw_set_local_storage_value(
        _get_persistent_path(gobj),
        attrs
    );
}

/************************************************************
 *
 ************************************************************/
function db_remove_persistent_attrs(
    gobj,
    keys  // str, list or dict.
)
{
    let jn_file = kw_get_local_storage_value(_get_persistent_path(gobj), null, false);

    let attrs = kw_clone_by_not_keys(
        gobj,
        jn_file,
        keys,
        false
    );

    kw_set_local_storage_value(
        _get_persistent_path(gobj),
        attrs
    );
}

/************************************************************
 *
 ************************************************************/
function db_list_persistent_attrs(gobj)
{
    // TODO
}

//=======================================================================
//      Expose the class via the global object
//=======================================================================
export {
    db_load_persistent_attrs,
    db_save_persistent_attrs,
    db_remove_persistent_attrs,
    db_list_persistent_attrs,
};
